src/core/executor.c

```diff
@@ -215,13 +243,68 @@ int main(int argc, char *argv[]) {
                 log_open();
         }
 
-        r = fdset_remove(fdset, fileno(arg_serialization));
-        if (r < 0)
-                return log_error_errno(r, "Failed to remove serialization fd from fd set: %m");
+        /* If we are forked directly then we will have all the required FDs already open, create a set to
+         * prepare for deserialization. If we are getting a socket instead we'll get them in an array via
+         * SCM_RIGHTS. */
+        if (arg_serialization) {
+                serialization = arg_serialization;
+
+                r = fdset_new_fill(/* filter_cloexec= */ 0, &fdset);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to create fd set: %m");
+        } else if (arg_manager_socket >= 0) {
+                _cleanup_close_ int pidfd = -EBADF;
+                _cleanup_free_ int *fds_array = NULL;
+                size_t n_fds_array = 0;
+
+                manager_socket = arg_manager_socket;
+
+                /* Wait for serialization FD from manager datagram socket */
+                r = receive_many_fds(manager_socket, &fds_array, &n_fds_array, 0);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to receive serialization FDs: %m");
+
+                /* There always at the very least the exec FD */
+                if (n_fds_array == 0)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                               "Received no FDs from manager socket.");
+
+                r = fdset_new_array(&fdset, fds_array, n_fds_array);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to create fd set: %m");
+
+                log_debug("Accepting work from worker " PID_FMT, getpid_cached());
+
+                /* Send back pidfd so that systemd knows who took up this job */
+                pidfd = pidfd_open(getpid_cached(), 0);
+                if (pidfd < 0) {
+                        char iov_buffer[DECIMAL_STR_MAX(pid_t) + 1] = {};
+                        struct iovec iov = IOVEC_MAKE(iov_buffer, sizeof(iov_buffer) - 1);
```

Comment:
this should use IOVEC_MAKE_STRING() or so, i.e. not send more byte sthan what we actually format into the string